src/main/java/htsjdk/variant/variantcontext/JEXLMap.java

```diff
@@ -11,152 +12,188 @@
+ * It lazily initializes each value as requested to save as much processing time as possible.
  *
+ * TODO: must resolve the following outdated comment in PR review rounds
  * Compatible with JEXL 1.1 (this code will be easier if we move to 2.0, all of the functionality can go into the
  * JexlContext's get()
  *
+ * TODO: there is some troubling design choices in this class: the laziness and exception behavior:
+ * TODO:   because the class is designed to be lazy, when an instance is created, some invalid expressions may be provided
+ * TODO:   later when user requests some/all values, those invalid expressions will result in IllegalArgumentException,
+ * TODO:   especially when calling the function values(), which is the last thing users expect when calling such accessors.
+ * TODO:   but there may be no good choice here. could decide to "unsupport" the values() function.
  */
 
 class JEXLMap implements Map<JexlVCMatchExp, Boolean> {
-    // our variant context and/or Genotype
-    private final VariantContext vc;
-    private final Genotype g;
 
     // our context
     private JexlContext jContext = null;
+    // our variant context and/or Genotype
+    private final VariantContext vc;
+    private final Genotype g;
 
-    // our mapping from JEXLVCMatchExp to Booleans, which will be set to NULL for previously uncached JexlVCMatchExp
+    /**
+     * our mapping from {@link JexlVCMatchExp} to {@link Boolean}s, which will be set to {@code NULL}
+     * for previously un-cached {@link JexlVCMatchExp}.
+     */
     private Map<JexlVCMatchExp,Boolean> jexl;
 
-
-    public JEXLMap(Collection<JexlVCMatchExp> jexlCollection, VariantContext vc, Genotype g) {
+    // -----------------------------------------------------------------------------------------------
+    // Initializers and accessors
+    // -----------------------------------------------------------------------------------------------
+    public JEXLMap(final Collection<JexlVCMatchExp> jexlCollection, final VariantContext vc, final Genotype g) {
+        lazyInitialize(jexlCollection);
         this.vc = vc;
         this.g = g;
-        initialize(jexlCollection);
     }
 
-    public JEXLMap(Collection<JexlVCMatchExp> jexlCollection, VariantContext vc) {
+    public JEXLMap(final Collection<JexlVCMatchExp> jexlCollection, final VariantContext vc) {
         this(jexlCollection, vc, null);
     }
 
-    private void initialize(Collection<JexlVCMatchExp> jexlCollection) {
-        jexl = new HashMap<JexlVCMatchExp,Boolean>();
-        for (JexlVCMatchExp exp: jexlCollection) {
-            jexl.put(exp, null);
-        }
-    }
-
     /**
-     * create the internal JexlContext, only when required.  This code is where new JEXL context variables
-     * should get added.
-     *
+     * Note: due to laziness, this accessor actually modifies the instance by possibly forcing evaluation of an Jexl expression.
+     * @throws IllegalArgumentException when {@code o} is {@code null}
```

Comment:
Does this also throw IllegalArgumentException in other cases like you mentioned in the todo?